### PR TITLE
licensing: enforce site admin roles

### LINKED ISSUE #14023: RFC 167: Enforce site-admin roles
The actual form of enforcement hasn't been decided yet, see discussions in the [RFC](https://docs.google.com/document/d/1XozQ4JINJqirdaG-XqGtboT2-PlIXPyBn6EwV7Q3pWI/edit?disco=AAAAJ_b2hIg).

Part of [RFC 167](https://docs.google.com/document/d/1XozQ4JINJqirdaG-XqGtboT2-PlIXPyBn6EwV7Q3pWI/edit).

TODO

- [x] All new users are prompted to be side admins in Free tier
- [x] Prevent user creation with a malformed/expired license
- [x] Dotcom: no new users can be prompted to site admin automatically
- [x] Prevent revoking site admins

### PR DESCRIPTION
**Pull request by unknwon**
Non-site admins are not allowed in Free tier.

- All new users are promoted to be side admins in Free tier
- Prevent user creation with a malformed/expired license
- Prevent revoking site admins
- Dotcom: no new users can be promoted to site admin automatically

Unit tests are added and tested end-to-end.

Fixes #14023

### CODE REVIEW

File: enterprise/cmd/frontend/internal/licensing/enforcement/users.go [hunk 1]
```diff
@@ -50,3 +58,55 @@ func NewPreCreateUserHook(s licensing.UsersStore) func(context.Context) error {
...
+			if err != nil {
+				return err
+			}
+		}
+
+		return nil
+	}
+}
+
+// NewPreSetUserIsSiteAdmin returns a PreSetUserIsSiteAdmin closure.
```
Comment (ryanslade) [inline enterprise/cmd/frontend/internal/licensing/enforcement/users.go:90]: It's not clear to me what this function is for

Comment (ryanslade) [reply enterprise/cmd/frontend/internal/licensing/enforcement/users.go:90]: Ah, I see lower down where the interface is defined. Might be useful adding more context here too that mentions when this function will be called

File: internal/db/users.go [hunk 2]
```diff
@@ -29,6 +30,13 @@ type users struct {
 	// PreCreateUser (if set) is a hook called before creating a new user in the DB by any means
 	// (e.g., both directly via Users.Create or via ExternalAccounts.CreateUserAndSave).
 	PreCreateUser func(context.Context) error
+	// AfterCreateUser (if set) is a hook called after creating a new user in the DB by any means
+	// (e.g., both directly via Users.Create or via ExternalAccounts.CreateUserAndSave).
+	// Whatever this hook mutates in database should be reflected on the `user` argument as well.
+	AfterCreateUser func(ctx context.Context, tx dbutil.DB, user *types.User) error
+	// PreSetUserIsSiteAdmin (if set) is a hook called before promoting/revoking a user to be a
+	// site admin.
+	PreSetUserIsSiteAdmin func(isSiteAdmin bool) error
```
Comment (ryanslade) [inline internal/db/users.go:39]: Maybe use `Before` instead of `Pre`?

I think it should be either `Pre` and `Post` or `Before` and `After`

**Review by ryanslade — APPROVED** [review]
LGTM apart from some potential renaming and additional comments

**Comment by unknwon** [comment]
@ryanslade Thanks! Addressed feedback in https://github.com/sourcegraph/sourcegraph/pull/15570/commits/49be0722049f746ea726d37bfe59501a3e14777a, merging now.

LMK if you have any post-merge comments.

**Pull request merged**